File: 02-Web开发/01-前端框架/WebComponents/web_components.go
// Custom Elements 最小实现的 Go 复刻，与 web_components.py 同题。
// 依据：WHATWG HTML「4.13 Custom elements」https://html.spec.whatwg.org/multipage/custom-elements.html
package main

import "fmt"

// 生命周期回调用匿名接口断言，未实现即跳过（等价于 Python 的 getattr(..., None)）
type connected interface{ ConnectedCallback() }
type disconnected interface{ DisconnectedCallback() }
type attrChanged interface{ AttributeChangedCallback(string, string, string) }
type connectedMove interface{ ConnectedMoveCallback() }
type observed interface{ ObservedAttributes() []string }

var upgradeStates = map[string]bool{"undefined": true, "uncustomized": true}

type Element struct {
	LocalName   string
	State       string
	IsConnected bool
	Attributes  map[string]string
	Children    []string
	Impl        interface{}
	Log         []string
}

func NewElement(localName string) *Element {
	return &Element{
		LocalName:  localName,
		State:      "undefined",
		Attributes: map[string]string{},
	}
}

func (e *Element) fire(name string) {
	e.Log = append(e.Log, name)
	switch c := e.Impl.(type) {
	case connected:
		if name == "connectedCallback" {
			c.ConnectedCallback()
		}
	case disconnected:
		if name == "disconnectedCallback" {
			c.DisconnectedCallback()
		}
	case connectedMove:
		if name == "connectedMoveCallback" {
			c.ConnectedMoveCallback()
		}
	}
}

type Registry struct {
	Definitions map[string]func(*Element) interface{}
	Queue       []func()
	Processing  bool
	Elements    []*Element
}

func NewRegistry() *Registry {
	return &Registry{Definitions: map[string]func(*Element) interface{}{}}
}

func (r *Registry) Enqueue(fn func()) { r.Queue = append(r.Queue, fn) }

func (r *Registry) Process() {
	if r.Processing {
		return
	}
	r.Processing = true
	for len(r.Queue) > 0 {
		fn := r.Queue[0]
		r.Queue = r.Queue[1:]
		fn()
	}
	r.Processing = false
}

func (r *Registry) Define(name string, ctor func(*Element) interface{}) {
	r.Definitions[name] = ctor
	for _, el := range r.Elements {
		e := el
		if e.LocalName == name && upgradeStates[e.State] && e.IsConnected {
			r.Enqueue(func() { r.UpgradeElement(e) })
		}
	}
}

// UpgradeElement 返回是否真的升级了：已是 custom/failed 的元素保持幂等。
func (r *Registry) UpgradeElement(el *Element) bool {
	ctor, ok := r.Definitions[el.LocalName]
	if !ok || !upgradeStates[el.State] {
		return false
	}
	// 构造期间 attributes/children 对元素不可见（规范强制）
	savedAttrs, savedChildren := el.Attributes, el.Children
	el.Attributes, el.Children = map[string]string{}, nil
	func() {
		defer func() {
			if rec := recover(); rec != nil {
				el.State = "failed"
			}
		}()
		el.Impl = ctor(el)
		el.State = "custom"
	}()
	el.Attributes, el.Children = savedAttrs, savedChildren
	if el.State != "custom" {
		return false
	}
	el.fire("constructor")
	if el.IsConnected {
		r.Enqueue(func() { el.fire("connectedCallback") })
	}
	return true
}

func (r *Registry) Upgrade(root *Element) int {
	targets := r.Elements
	if root != nil {
		targets = []*Element{root}
	}
	done := 0
	for _, el := range targets {
		e := el
		if _, ok := r.Definitions[e.LocalName]; ok && upgradeStates[e.State] {
			r.Enqueue(func() { r.UpgradeElement(e) })
			done++
		}
	}
	return done
}

func (r *Registry) Connect(el *Element) {
	el.IsConnected = true
	if _, ok := r.Definitions[el.LocalName]; ok && upgradeStates[el.State] {
		r.Enqueue(func() { r.UpgradeElement(el) })
		return
	}
	if el.State == "custom" {
		r.Enqueue(func() { el.fire("connectedCallback") })
	}
}

func (r *Registry) Disconnect(el *Element) {
	el.IsConnected = false
	if el.State == "custom" {
		r.Enqueue(func() { el.fire("disconnectedCallback") })
	}
}

func (r *Registry) Move(el *Element) {
	if _, ok := el.Impl.(connectedMove); ok {
		r.Enqueue(func() { el.fire("connectedMoveCallback") })
		return
	}
	r.Enqueue(func() { el.fire("disconnectedCallback") })
	r.Enqueue(func() { el.fire("connectedCallback") })
}

func (r *Registry) SetAttribute(el *Element, name, value string) {
	old := el.Attributes[name]
	el.Attributes[name] = value
	if el.State != "custom" {
		return
	}
	if o, ok := el.Impl.(observed); ok {
		for _, n := range o.ObservedAttributes() {
			if n == name {
				on, ov := name, old
				r.Enqueue(func() {
					if c, ok2 := el.Impl.(attrChanged); ok2 {
						el.Log = append(el.Log, "attributeChangedCallback")
						c.AttributeChangedCallback(on, ov, value)
					}
				})
			}
		}
	}
}

// ---------- 示例实现 ----------

type widget struct {
	el    *Element
	conns []bool
}

func (w *widget) ObservedAttributes() []string { return []string{"title"} }
func (w *widget) ConnectedCallback()           { w.conns = append(w.conns, w.el.IsConnected) }

func main() {
	r := NewRegistry()
	el := NewElement("my-widget")
	el.IsConnected = true
	r.Elements = []*Element{el}
	r.Define("my-widget", func(e *Element) interface{} { return &widget{el: e} })
	r.Process()
	fmt.Println("after define + process:", el.State, el.Log)

	r.Disconnect(el)
	r.Process()
	r.Connect(el)
	r.Process()
	fmt.Println("reconnect:", el.Log)

	detached := NewElement("my-widget")
	n := r.Upgrade(detached)
	r.Process()
	fmt.Println("manual upgrade:", detached.State, detached.Log, "queued:", n)
}

File: 02-Web开发/01-前端框架/WebComponents/web_components.py
# -*- coding: utf-8 -*-
"""
Custom Elements 最小实现：升级（upgrade）+ custom element reactions 队列。

权威依据：WHATWG HTML「4.13 Custom elements」
  https://html.spec.whatwg.org/multipage/custom-elements.html
  - §4.13.5 Upgrades：upgrade 只作用于 state 为 "undefined"/"uncustomized" 的元素；
    未插入文档的元素不会被自动升级（"upgrades only apply to elements in the document"）。
  - §4.13.2/§4.13.6：构造函数里**不得**检查 attributes/children —— non-upgrade 情形下它们
    根本不存在；也不得新增 attributes/children。工作应尽量推迟到 connectedCallback，
    "note that connectedCallback can be called more than once"。
  - reaction 队列：connectedCallback 可以在元素已 disconnected 之后仍被调用，
    此时回调里 `isConnected` 为 false（规范给了这段示例输出）。
  - §4.13.2.1 移动元素：默认会依次调用 disconnectedCallback 与 connectedCallback；
    定义了 connectedMoveCallback 则取代默认行为。
  - CustomElementRegistry.upgrade(root)：手动升级尚不在文档中的元素。
  - attributeChangedCallback 只对 observedAttributes 里的属性触发。
"""

UPGRADE_STATES = ("undefined", "uncustomized")


class Element:
    def __init__(self, local_name, impl_factory=None):
        self.local_name = local_name
        self.state = "undefined"          # undefined/uncustomized/custom/failed
        self.is_connected = False
        self.attributes = {}
        self.children = []
        self.impl = None
        self.log = []
        self.factory = impl_factory
        self.doc = None

    # 生命周期回调（由 impl 可选实现）
    def _fire(self, name, *args):
        self.log.append(name)
        fn = getattr(self.impl, name, None)
        if fn is not None:
            fn(*args)


class CustomElementRegistry:
    def __init__(self):
        self.definitions = {}             # localName -> 可调用（构造函数）
        self.queue = []                   # custom element reaction queue
        self.processing = False

    # ---------- 队列 ----------
    def enqueue(self, reaction):
        self.queue.append(reaction)

    def process(self):
        """处理 reaction 队列；处理期间新入队的 reaction 追加到同一轮尾部。"""
        if self.processing:
            return
        self.processing = True
        while self.queue:
            fn = self.queue.pop(0)
            fn()
        self.processing = False

    def define(self, name, ctor):
        self.definitions[name] = ctor
        # 已定义后，文档中处于 undefined 状态的元素会被升级
        for el in list(getattr(self, "_elements", [])):
            if el.local_name == name and el.state in UPGRADE_STATES and el.is_connected:
                self.enqueue(lambda e=el: self.upgrade_element(e))
        return self

    def track(self, elements):
        self._elements = list(elements)
        return self

    # ---------- 升级 ----------
    def upgrade_element(self, el):
        ctor = self.definitions.get(el.local_name)
        if ctor is None:
            return False
        if el.state not in UPGRADE_STATES:
            return False                  # 已经升级过 / 已失败：幂等
        saved_attrs, saved_children = el.attributes, el.children
        # 构造期间 attributes/children 对元素不可见（规范强制）
        el.attributes, el.children = {}, []
        try:
            el.impl = ctor(el)
            el.state = "custom"
        except Exception:
            el.state = "failed"
            el.attributes, el.children = saved_attrs, saved_children
            return False
        el.attributes, el.children = saved_attrs, saved_children
        el._fire("constructor")
        if el.is_connected:
            # 升级一个已连接的元素时，connectedCallback 紧跟在构造函数之后
            self.enqueue(lambda: el._fire("connectedCallback"))
        return True

    def upgrade(self, root=None):
        """registry.upgrade(root)：手动升级（不要求元素已在文档中）。"""
        targets = [root] if root is not None else list(getattr(self, "_elements", []))
        done = 0
        for el in targets:
            if el.local_name in self.definitions and el.state in UPGRADE_STATES:
                self.enqueue(lambda e=el: self.upgrade_element(e))
                done += 1
        return done

    # ---------- 树操作 ----------
    def connect(self, el):
        el.is_connected = True
        if el.local_name in self.definitions and el.state in UPGRADE_STATES:
            # 升级时由 upgrade_element 负责补一次 connectedCallback
            self.enqueue(lambda: self.upgrade_element(el))
        elif el.state == "custom":
            # 已经是自定义元素（例如断开后重连）→ 第二次 connectedCallback
            self.enqueue(lambda: el._fire("connectedCallback"))
        # 未定义的名字（普通 <div>）没有任何 custom element reaction

    def disconnect(self, el):
        el.is_connected = False
        if el.state == "custom":
            self.enqueue(lambda: el._fire("disconnectedCallback"))

    def move(self, el):
        """移动元素：默认 disconnected + connected，除非实现了 connectedMoveCallback。"""
        if hasattr(el.impl or object(), "connectedMoveCallback"):
            self.enqueue(lambda: el._fire("connectedMoveCallback"))
        else:
            self.enqueue(lambda: el._fire("disconnectedCallback"))
            self.enqueue(lambda: el._fire("connectedCallback"))

    def set_attribute(self, el, name, value):
        old = el.attributes.get(name)
        el.attributes[name] = value
        if el.state == "custom" and name in getattr(el.impl, "observedAttributes", []):
            self.enqueue(lambda: el._fire("attributeChangedCallback", name, old, value))
        return old

    def parse_children(self, el, children):
        """模拟解析器填充子节点（构造之后才发生）。"""
        el.children = list(children)
